### PR TITLE
Added option to get resolved packages in a requirements.txt file. Fixes #135

### LINKED ISSUE #135: Add option to get resolved packages in a requirements.txt file

### PR DESCRIPTION
**Pull request by arijitde92**
Created a new option `--resolved-output` that finds resolved packages and writes those into the given filename in a typical requirements.txt file format. Also added test test_resolved_cli in test_cli.py.

### CODE REVIEW

**Comment by arijitde92** [comment]
I was having an issue rebasing in the previous PR #156 . Hence opened this new PR. It is working as expected. Please check.

**Comment by TG1999** [comment]
@arijitde92 tests are failing in CI, please check. Thanks!

**Comment by arijitde92** [comment]
Hi @TG1999 ,

After inspecting the cause of test failure, I found that a few test cases are failing but those test cases are not written by me. Please see screenshots below-
macos13
ubuntu22
win2022

As you  can see, the same tests are failing for all the OSes and none of them are written by me. Is this happening with other PRs too? Or is this happening somehow because of the code I wrote? As I am not able to find any relation between the code I wrote and the tests that are failing.

Please guide me on this.
Thanks.

**Comment by TG1999** [comment]
@arijitde92 please take a pull from latest main branch, main branch is green :)

**Comment by arijitde92** [comment]
Hi @TG1999 , I have pulled the latest main branch in my forked repo as you can see the latest commit in the screenshot below.
image


But still 1 test is failing in CI as shown in below screenshot.
image

**Comment by TG1999** [comment]
You have to run 'make valid' command to fix formatting errors

**Comment by arijitde92** [comment]
Hi @TG1999 , checks are now passing. Let me know if anything else is required.

File: src/python_inspector/utils.py [hunk 1]
```diff
@@ -54,6 +54,20 @@ def write_output_in_file(output, location):
     return output
 
 
+def write_resolved_packages(package_list, requirements_file):
```
Comment (TG1999) [inline src/python_inspector/utils.py:57]: @arijitde92 please add a separate unit test for this function, separated from CLI logic

File: src/python_inspector/utils.py [hunk 2]
```diff
@@ -54,6 +54,20 @@ def write_output_in_file(output, location):
...
+    """
+    Write the resolved package names and versions into ``requirements_file_path``
+    """
+    dependencies = package_list[0]["package_data"][0]["dependencies"]
```
Comment (TG1999) [inline src/python_inspector/utils.py:61]: what if `package_list` is empty? and `package_data` does not exist there

File: src/python_inspector/utils.py [hunk 3]
```diff
@@ -54,6 +54,20 @@ def write_output_in_file(output, location):
...
+    resolved_packages = []
+    for dependency in dependencies:
+        if dependency["is_resolved"]:
```
Comment (TG1999) [inline src/python_inspector/utils.py:64]:
```suggestion
        if dependency.get("is_resolved"):
```

File: src/python_inspector/utils.py [hunk 4]
```diff
@@ -54,6 +54,20 @@ def write_output_in_file(output, location):
...
+            package = dependency["extracted_requirement"]
```
Comment (TG1999) [inline src/python_inspector/utils.py:65]:
```suggestion
            package = dependency.get("extracted_requirement")
```

File: src/python_inspector/utils.py [hunk 5]
```diff
@@ -54,6 +54,20 @@ def write_output_in_file(output, location):
...
+            resolved_packages.append(package)
```
Comment (TG1999) [inline src/python_inspector/utils.py:66]:
```suggestion
            if package:
                resolved_packages.append(package)
```

**Review by TG1999 — CHANGES_REQUESTED** [review]
@arijitde92 thanks++, some nits for your consideration

File: src/python_inspector/utils.py (hunk 1 above)
Comment (arijitde92) [reply src/python_inspector/utils.py:57]: @TG1999 , I have written an unit test in test_cli.py (See [here](https://github.com/arijitde92/python-inspector/blob/235a0563699e3d718db874f3819a5f4cc976784f/tests/test_cli.py#L448)). Do I need to write another unit test elsewhere (maybe in test_utils.py?)?
Also could you please clarify what "separated from CLI logic" means?
I assume that you mean to test for a valid `package_list`. In that case do I need to create a dummy `package_list`?

Comment (arijitde92) [reply src/python_inspector/utils.py:57]: Hi @TG1999 , could you please help in clarifying what kind of unit test you need? And in which while should I write the test?

Comment (TG1999) [reply src/python_inspector/utils.py:57]: @arijitde92 yes, please write a test in `test_utils.py`

File: src/python_inspector/utils.py (hunk 2 above)
Comment (arijitde92) [reply src/python_inspector/utils.py]: I have implemented the checks for empty or corrupt `package_list`.

File: src/python_inspector/utils.py (hunk 1 above)
Comment (arijitde92) [reply src/python_inspector/utils.py:57]: I have written two unit tests in test_utils.py. Please check.

**Comment by arijitde92** [comment]
Hi @TG1999 , I have added unit tests as requested and also edited the code as requested.
But now some tests are failing again. Please check in the below screenshot-
image

**Comment by pombredanne** [comment]
@arijitde92 too bad you deleted your repo!

**Pull request closed without merging**